x-pack/plugins/endpoint/server/routes/whitelist.ts

```diff
@@ -0,0 +1,346 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { IRouter, RequestHandlerContext, APICaller } from 'kibana/server';
+import { schema } from '@kbn/config-schema';
+import { SearchResponse } from 'elasticsearch';
+import { createHash } from 'crypto';
+import { i18n } from '@kbn/i18n';
+
+import lzma from 'lzma-native';
+import { WhitelistRule, WhitelistSet } from '../../common/types';
+import { EndpointAppContext } from '../types';
+
+const whitelistIdx: string = 'whitelist'; // TODO: change this
+const allowlistBaseRoute: string = '/api/endpoint/allowlist';
+const whitelistLimit: number = 1000;
+let whitelistArtifactCache: Buffer = Buffer.from([]);
+let whitelistCount: number = 0;
+
+/**
+ * Registers the whitelist routes for the API
+ */
+export function registerWhitelistRoutes(router: IRouter, endpointAppContext: EndpointAppContext) {
+  router.get(
+    {
+      path: allowlistBaseRoute,
+      validate: {},
+      options: { authRequired: true },
+    },
+    handleWhitelistGet
+  );
+
+  router.get(
+    {
+      path: '/api/endpoint/manifest',
+      validate: {},
+      options: { authRequired: true },
+    },
+    handleWhitelistManifest
+  );
+  router.get(
+    {
+      path: `${allowlistBaseRoute}/download/{hash}`,
+      validate: {
+        params: schema.object({
+          hash: schema.string(),
+        }),
+      },
+      options: { authRequired: true },
+    },
+    handleWhitelistDownload
+  );
+
+  router.post(
+    {
+      path: allowlistBaseRoute,
+      validate: {
+        body: schema.object({
+          comment: schema.maybe(schema.string()), // Optional comment explaining reason for whitelist
+          event_types: schema.arrayOf(schema.string()),
+          file_path: schema.maybe(schema.string()),
+          signer: schema.maybe(schema.string()),
+          sha256: schema.maybe(schema.string()),
+          dismiss: schema.boolean({ defaultValue: false }), // Boolean determining if we dismiss all alerts that match this
+        }),
+      },
+      options: { authRequired: true },
+    },
+    handleWhitelistPost
+  );
+
+  router.delete(
+    {
+      path: allowlistBaseRoute,
+      validate: {
+        body: schema.object({
+          whitelist_id: schema.string(),
+        }),
+      },
+      options: { authRequired: true },
+    },
+    handleWhitelistItemDeletion
+  );
+}
+
+/**
+ * Handles the POST request for whitelist additions
+ */
+async function handleWhitelistPost(context: RequestHandlerContext, req, res) {
+  try {
+    const eventTypes: string[] = req.body.eventTypes;
+    const whitelistAttributeMap: Record<string, string> = {
+      file_path: 'malware.file.path',
+      acting_process_path: 'actingProcess.file.path',
+      signer: 'malware.file.signer',
+      sha256: 'malware.file.hashes.sha256',
+    };
+    const newRules: WhitelistRule[] = [];
+    Object.keys(whitelistAttributeMap).forEach(k => {
+      if (req.body[k]) {
```

Comment:
Do we ever have a case where we'd want to match on a `null` or `undefined` value? Because if so, this will skip those.

Reply:
I can't imagine a case where we'd want to whitelist based on empty values